### ISSUE TITLE
WriteDummyDBF in WriteGeometryCollection Method

### ISSUE THREAD

**Issue by petlof** [NetTopologySuite/NetTopologySuite#56]
Having some problems with my DBF-file is getting it's header overwritten when writing the collection of shapes to disk. 
Found that it's actually a line in the WriteGeometryCollection-method that overwrites my DBF-file that I already have created.

I don't like this design since it's not clear for the one using NTS that the DBF-file needs to be written after the shape-file.
Shouldn't we add a check here and not create the DummyDbf (if it should be there at all) if the DBF-header of DbfWriter already have been written. Of if the DBF-file already exists.

https://github.com/NetTopologySuite/NetTopologySuite/blob/master/NetTopologySuite.IO/NetTopologySuite.IO.GeoTools/ShapefileWriter.cs#L188

https://github.com/NetTopologySuite/NetTopologySuite/blob/master/NetTopologySuite.IO/NetTopologySuite.IO.GeoTools/ShapefileWriter.cs#L204

**Comment by petlof** [comment]
jaundice commented 3 minutes ago
I don't really like the dummy dbf writing, but it predates my changes and I didn't have time to restructure it.

**Comment by petlof** [comment]
Ok, I don't know what we should do with it, but it feels that we at least should make sure we don't overwrite an existing dbf-file. 
We could probably expose the HeaderWritten on the DBFFileHeader and skip the dummy-writing if the header have been written already. Is that an OK solution?

**Comment by jaundice** [comment]
Personally I think it should either write both shape and attribute data in the same method, or shape and dummy dbf, however other people may rely on the api as is.

**Comment by petlof** [comment]
I talked to @FObermaier and we said we could a a method on shapefilewirter for WriteFeatures(IEnumerable<IFeature> features);

Then one would not need to handle the specific writers individually..
I will look into that more tomorrow..

This would also enable things like streaming big datasets to shapefile without having to store all rows in RAM

**Comment by jaundice** [comment]
That is a far more elegant solution

**Comment by petlof** [comment]
Suggestion on method for writing features as part of the ShapefileWriter class..

Does it look OK @jaundice @FObermaier ?, in that case I can commit it.
==
/// <summary>
        /// Write the enumeration of features to shapefile (shp + dbf)
        /// </summary>
        /// <param name="filename">Filename to create</param>
        /// <param name="features">Enumeration of features to write, features will be enumerated once</param>
        /// <param name="fields">Fields that should be written, only those attributes specified here will be mapped from the feature attributetable while writing</param>
        /// <param name="shapeGeometryType">Type of geometries shapefile</param>
        /// <param name="dbfEncoding">Optional Encoding to be used when writing the DBF-file (default Windows-1252)</param>
        public static void WriteFeatures(string filename, IEnumerable<IFeature> features, DbaseFieldDescriptor[] fields, ShapeGeometryType shapeGeometryType, 
            Encoding dbfEncoding = null)
        {
            
                if (dbfEncoding == null)
                dbfEncoding = Encoding.GetEncoding(1252);
            
                using (var shpWriter = new ShapefileWriter(Path.ChangeExtension(filename,".shp"), shapeGeometryType))
                {
                    using (var dbfWriter = new DbaseFileWriter(Path.ChangeExtension(filename, ".dbf"), dbfEncoding))
                    {
                        var dbfHeader = new DbaseFileHeader(dbfEncoding);
                        foreach (var field in fields)
                        {
                            dbfHeader.AddColumn(field.Name, field.DbaseType, field.Length, field.DecimalCount);
                        }
                        dbfWriter.Write(dbfHeader);

                        int numFeatures = 0;
                        foreach (var feature in features)
                        {
                            shpWriter.Write(feature.Geometry);
                            var values = new object[fields.Length];
                            for (int i = 0; i < fields.Length; i++)
                            {
                                values[i] = feature.Attributes[fields[i].Name];
                            }
                            dbfWriter.Write(values);
                            numFeatures++;
                        }

                        dbfHeader.NumRecords = numFeatures;
                        dbfWriter.Write(dbfHeader);
                    }
                }
        }
==

**Comment by petlof** [comment]
Small change in DBaseFileWriter also that rewinds the _writer when header is written to be able to write it once we're done

```
       /// <summary>
        /// </summary>
        /// <param name="header"></param>
        public void Write(DbaseFileHeader header)
        {
            if (header == null)
                throw new ArgumentNullException("header");
            //if (_recordsWritten)
            //    throw new InvalidOperationException("Records have already been written. Header file needs to be written first.");
            _headerWritten = true;

            if (header.Encoding.WindowsCodePage != _encoding.WindowsCodePage)
            {
                header.Encoding = _encoding;
            }

            //Header should always be written first in the file
            if (_writer.BaseStream.Position != 0)
                _writer.Seek(0, SeekOrigin.Begin);

            header.WriteHeader(_writer);
            _header = header;
        }
```

**Comment by jaundice** [comment]
Looks good to me Peter..

**Comment by jaundice** [comment]
I haven’t looked at the wider code for this, but should we consider caching the position in the stream before we write the header and moving back to it after the header is written?


Small change in DBaseFileWriter also that rewinds the _writer when header is written to be able to write it once we're done

       /// <summary>

        /// </summary>

        /// <param name="header"></param>

        public void Write(DbaseFileHeader header)

        {

            if (header == null)

                throw new ArgumentNullException("header");

            //if (_recordsWritten)

            //    throw new InvalidOperationException("Records have already been written. Header file needs to be written first.");

            _headerWritten = true;



            if (header.Encoding.WindowsCodePage != _encoding.WindowsCodePage)

            {

                header.Encoding = _encoding;

            }



            //Header should always be written first in the file

            if (_writer.BaseStream.Position != 0)

                _writer.Seek(0, SeekOrigin.Begin);



            header.WriteHeader(_writer);

            _header = header;

        }

**Comment by petlof** [comment]
Sure, I can do that..

**Comment by FObermaier** [comment]
@petlof, @jaundice:
I've submitted @petlof's code including @jaundice's suggestion to expose _headerWritten.
In WriteDummyDbf there was a call to _writer.Write(0x1A) (=> _writer.WriteEndOfDbf()). I've added that to WriteFeatures and modified the DbaseFileWriter.Write(DbaseHeader) function to reposition the stream accordingly.

Could you please verify.

**Comment by petlof** [comment]
Looks good!

**Comment by FObermaier** [comment]
Fixed with 69e8ddbc21c3efc381d05aa84b3b573638f88146
